ratbagd/ratbagd-button.c

```diff
@@ -266,28 +278,79 @@ static int ratbagd_button_set_macro(sd_bus *bus,
 	return 0;
 }
 
-static int ratbagd_button_get_action_type(sd_bus *bus,
-					  const char *path,
-					  const char *interface,
-					  const char *property,
-					  sd_bus_message *reply,
-					  void *userdata,
-					  sd_bus_error *error)
+static int ratbagd_button_get_mapping(sd_bus *bus,
+				      const char *path,
+				      const char *interface,
+				      const char *property,
+				      sd_bus_message *reply,
+				      void *userdata,
+				      sd_bus_error *error)
 {
 	struct ratbagd_button *button = userdata;
 	enum ratbag_button_action_type type;
 
 	type = ratbag_button_get_action_type(button->lib_button);
 	if (type == RATBAG_BUTTON_ACTION_TYPE_KEY)
 		type = RATBAG_BUTTON_ACTION_TYPE_UNKNOWN;
-
 	verify_unsigned_int(type);
 
-	CHECK_CALL(sd_bus_message_append(reply, "u", type));
+	switch (type) {
+	case RATBAG_BUTTON_ACTION_TYPE_BUTTON:
+		return ratbagd_button_get_button(bus, path, interface, property,
+						 reply, userdata, error);
+	case RATBAG_BUTTON_ACTION_TYPE_SPECIAL:
+		return ratbagd_button_get_special(bus, path, interface, property,
+						 reply, userdata, error);
+	case RATBAG_BUTTON_ACTION_TYPE_MACRO:
+		return ratbagd_button_get_macro(bus, path, interface, property,
+						reply, userdata, error);
+	default:
+		return sd_bus_message_append(reply, "(uv)",
+					     RATBAG_BUTTON_ACTION_TYPE_UNKNOWN,
+					     "u", 0);
+	}
 
 	return 0;
 }
 
+static int ratbagd_button_set_mapping(sd_bus *bus,
+				      const char *path,
+				      const char *interface,
+				      const char *property,
+				      sd_bus_message *m,
+				      void *userdata,
+				      sd_bus_error *error)
+{
+	enum ratbag_button_action_type type;
+
+	CHECK_CALL(sd_bus_message_enter_container(m, SD_BUS_TYPE_STRUCT, "uv"));
+	CHECK_CALL(sd_bus_message_read(m, "u", &type));
+
+	switch (type) {
+	case RATBAG_BUTTON_ACTION_TYPE_BUTTON:
+		CHECK_CALL(ratbagd_button_set_button(bus, path, interface, property,
+						     m, userdata, error));
+		break;;
+	case RATBAG_BUTTON_ACTION_TYPE_SPECIAL:
+		CHECK_CALL(ratbagd_button_set_special(bus, path, interface, property,
+						      m, userdata, error));
+		break;
+	case RATBAG_BUTTON_ACTION_TYPE_MACRO:
+		CHECK_CALL(ratbagd_button_set_macro(bus, path, interface, property,
+						    m, userdata, error));
+		break;
+	default:
+		/* FIXME */
```

Comment:
What is the intended fix here? Log a warning, silently ignore, return a special value to the caller?

Reply:
huh, good point, I forgot about that `abort()` in there. Removed it now with a silent `return 1;`, it wouldn't be good if a client can make `ratbagd` abort on command.

As for error handling in general, this needs to be specified in more detail anyway, it's very inconsistent right now.